### PR TITLE
feat: Customize node groups and user-data

### PR DESCRIPTION
**Pull request by ichaban-pinterest**
## Summary
Added several customizations for:
- `user-data` module and templates: added support for `format_mount_nvme_disk`, `kubelet_extra_args` for Linux template only
- Managed Node Groups: added support for `format_mount_nvme_disk`, `kubelet_extra_args`; removed `ignore_changes` for `desired_size` from scaling config
- Self-Managed Node Groups: added support for `kubelet_extra_args`, `labels`

### CODE REVIEW

File: modules/_user_data/README.md [hunk 1]
```diff
@@ -37,9 +37,12 @@ No modules.
 | <a name="input_cluster_endpoint"></a> [cluster\_endpoint](#input\_cluster\_endpoint) | Endpoint of associated EKS cluster | `string` | `""` | no |
 | <a name="input_cluster_name"></a> [cluster\_name](#input\_cluster\_name) | Name of the EKS cluster | `string` | `""` | no |
 | <a name="input_cluster_service_ipv4_cidr"></a> [cluster\_service\_ipv4\_cidr](#input\_cluster\_service\_ipv4\_cidr) | The CIDR block to assign Kubernetes service IP addresses from. If you don't specify a block, Kubernetes assigns addresses from either the 10.100.0.0/16 or 172.20.0.0/16 CIDR blocks | `string` | `null` | no |
+| <a name="input_cluster_service_ipv6_cidr"></a> [cluster\_service\_ipv6\_cidr](#input\_cluster\_service\_ipv4\_cidr) | The CIDR block to assign Kubernetes pod and service IP addresses from if `ipv6` was specified when the cluster was created. Kubernetes assigns service addresses from the unique local address range (fc00::/7) because you can't specify a custom IPv6 CIDR block when you create the cluster | `string` | `null` | no |
```
Comment (sfbahr) [inline modules/_user_data/README.md]: Looks like this is potentially conflicting with this PR upstream: https://github.com/terraform-aws-modules/terraform-aws-eks/pull/2960/files#diff-74a7b39721d8327e70e4d5a6de0affd3d1763389f1e570e51b1dbd60588815e5R42

File: templates/bottlerocket_user_data.tpl [hunk 2]
```diff
@@ -1,7 +1,8 @@
 %{ if enable_bootstrap_user_data ~}
+${pre_bootstrap_user_data ~}
 [settings.kubernetes]
-"cluster-name" = "${cluster_name}"
 "api-server" = "${cluster_endpoint}"
 "cluster-certificate" = "${cluster_auth_base64}"
+"cluster-name" = "${cluster_name}"
+${post_bootstrap_user_data ~}
```
Comment (sfbahr) [inline templates/bottlerocket_user_data.tpl]: I'm curious why these two changes were necessary?

File: modules/eks-managed-node-group/main.tf [hunk 3]
```diff
@@ -379,9 +382,6 @@ resource "aws_eks_node_group" "this" {
 
   lifecycle {
     create_before_destroy = true
-    ignore_changes = [
-      scaling_config[0].desired_size,
```
Comment (sfbahr) [inline modules/eks-managed-node-group/main.tf:383]: Could we describe why this change was needed?

File: modules/self-managed-node-group/main.tf [hunk 4]
```diff
@@ -144,8 +149,20 @@ resource "aws_launch_template" "this" {
...
+      arn = var.create_iam_instance_profile ? aws_iam_instance_profile.this[0].arn : var.iam_instance_profile_arn
+    }
+  }
+
+  dynamic "iam_instance_profile" {
+    for_each = !var.create_iam_instance_profile && var.iam_role_name != null ? [1] : []
+
+    content {
+      name = var.iam_role_name
+    }
```
Comment (sfbahr) [inline modules/self-managed-node-group/main.tf]: I'd like to see a rationale for this change as well

**Review by sfbahr — APPROVED** [review]
These changes generally seem reasonable. I'd like to see a bit more of the background behind these changes from our internal PRs brought out here. I'll do a bit of digging myself.

We should think about how we will merge our `ipv6_cidr` variable with the upstream PR for when we upgrade the base of our fork.

File: modules/self-managed-node-group/main.tf (hunk 4 above)
Comment (sfbahr) [reply modules/self-managed-node-group/main.tf]: From our internal PR:
>allow passing along IAM roles for node group instead of creating a new one

File: modules/eks-managed-node-group/main.tf (hunk 3 above)
Comment (ichaban-pinterest) [reply modules/eks-managed-node-group/main.tf:383]: We encountered an issue before where `desired_size` changes in the code were ignored and had no effect, so it was impossible to change the `desired_size` for a node group from code.

Related upstream issue https://github.com/terraform-aws-modules/terraform-aws-eks/issues/835
Related internal ticket is SRE-14776, you can find more context and discussions there.

File: templates/bottlerocket_user_data.tpl (hunk 2 above)
Comment (ichaban-pinterest) [reply templates/bottlerocket_user_data.tpl]: We don't really use Bottlerocket, it was mostly done for consistency with the Linux template we use.
It's not necessary and can be removed to minimize differences compared to the upstream.

File: modules/self-managed-node-group/main.tf (hunk 4 above)
Comment (ichaban-pinterest) [reply modules/self-managed-node-group/main.tf]: I don't have much context on this particular change since it was done by another folk.
This change was picked up from our internal repo, namely:
> Improve support for self-managed node groups `#29`

I will circle back with the author to provide more information on this.

File: modules/_user_data/README.md (hunk 1 above)
Comment (ichaban-pinterest) [reply modules/_user_data/README.md]: It was not implemented in v19.17.2 and there was an ask to add this feature.
But, AFAIK, we don't use it so we can drop it to minimize differences with the upstream and ease the future upgrade process.

File: templates/bottlerocket_user_data.tpl (hunk 2 above)
Comment (ichaban-pinterest) [reply templates/bottlerocket_user_data.tpl]: Removed in https://github.com/pinterest/terraform-aws-eks/pull/3/commits/e3413695d157974289581b7434220055c456f8f5

File: modules/_user_data/README.md (hunk 1 above)
Comment (ichaban-pinterest) [reply modules/_user_data/README.md]: Removed in https://github.com/pinterest/terraform-aws-eks/pull/3/commits/809061c0ed1fa327001402cb8254bb3e536ecb59

File: modules/self-managed-node-group/main.tf (hunk 4 above)
Comment (sfbahr) [reply modules/self-managed-node-group/main.tf]: From what I can see in the code here, I think we assume an instance profile already exists whose name matches the role the instance profile is using.

I think we could end up in an error condition (two iam_instance_profile blocks) if both `var.iam_instance_profile_arn` and `var.iam_role_name` are set with `var.create_iam_instance_profile == false`

File: modules/eks-managed-node-group/main.tf (hunk 3 above)
Comment (sfbahr) [reply modules/eks-managed-node-group/main.tf:383]: We should mention this change explicitly, since it's now inconsistent with the FAQ docs: https://github.com/pinterest/terraform-aws-eks/blob/pinterest-release-19.17.2/docs/faq.md#why-are-there-no-changes-when-a-node-groups-desired_size-is-modified

File: variables.tf [hunk 5]
```diff
@@ -86,12 +86,6 @@ variable "cluster_service_ipv4_cidr" {
   default     = null
 }
 
-variable "cluster_service_ipv6_cidr" {
```
Comment (sfbahr) [inline variables.tf]: Let's not deviate from upstream here, keep this variable and usage through the root module

File: modules/self-managed-node-group/main.tf (hunk 4 above)
Comment (sfbahr) [reply modules/self-managed-node-group/main.tf]: I think fine to keep this as is, but we should probably create a followup PR to handle this case ^

File: variables.tf (hunk 5 above)
Comment (ichaban-pinterest) [reply variables.tf]: Sure, reverted only those additional variables, kept the original ones in https://github.com/pinterest/terraform-aws-eks/commit/3e8b34ef1c404980e185aef006de3d2f264ec8a8

**Review by sfbahr — APPROVED** [review]
This changeset looks good to me, given:
1. We followup on the instance profile possible error condition: https://github.com/pinterest/terraform-aws-eks/pull/3/files#r1684539695
2. We clearly document that we are manually sizing the managed node groups and autoscaling won't work as expected.

File: modules/eks-managed-node-group/main.tf (hunk 3 above)
Comment (ichaban-pinterest) [reply modules/eks-managed-node-group/main.tf:383]: Updated the docs in https://github.com/pinterest/terraform-aws-eks/commit/4348e87708e4a4b4dfbb04e4082fd0734b3fdbe3

File: modules/self-managed-node-group/main.tf (hunk 4 above)
Comment (ichaban-pinterest) [reply modules/self-managed-node-group/main.tf]: Agree, it's fine for now and we can revise this later to eliminate possible conflict.

Comment (ichaban-pinterest) [reply modules/self-managed-node-group/main.tf]: Fixed the logic in https://github.com/pinterest/terraform-aws-eks/pull/3/commits/dc33cfcfef2a996fd41ade56cf78571011e0f313

**Comment by ichaban-pinterest** [comment]
Addressed both concerns:
1. Fixed the logic in https://github.com/pinterest/terraform-aws-eks/commit/dc33cfcfef2a996fd41ade56cf78571011e0f313
2. Updated the docs in https://github.com/pinterest/terraform-aws-eks/commit/4348e87708e4a4b4dfbb04e4082fd0734b3fdbe3

File: modules/self-managed-node-group/main.tf [hunk 6]
```diff
@@ -144,8 +149,20 @@ resource "aws_launch_template" "this" {
...
+      arn = var.iam_instance_profile_arn
+    }
+  }
+
+  dynamic "iam_instance_profile" {
+    for_each = !var.create_iam_instance_profile && var.iam_instance_profile_arn == null && var.iam_role_name != null ? [1] : []
+
+    content {
+      name = var.iam_role_name
+    }
```
Comment (sfbahr) [inline modules/self-managed-node-group/main.tf]: This change prevents the instance profiled created w/ `var.create_iam_instance_profile` from being used.

Since we're fixing this here, I think this code is overly complicated so here's my suggestion:
```suggestion
  iam_instance_profile {
    arn = try(aws_iam_instance_profile.this[0].arn, var.iam_instance_profile_arn)
    name = try(aws_iam_instance_profile.this[0].arn, var.iam_instance_profile_arn) == null ? var.iam_role_name : null
  }
```

**Review by sfbahr — CHANGES_REQUESTED** [review]

File: modules/self-managed-node-group/main.tf (hunk 6 above)
Comment (ichaban-pinterest) [reply modules/self-managed-node-group/main.tf]: Thanks for the suggestion, it seems reasonable but it doesn't utilize `var.create_iam_instance_profile` like the original code does. I'm wondering if it could lead to some potential issues, or just `try` among two options for both `arn` and `name` is enough to cover all cases.

Comment (sfbahr) [reply modules/self-managed-node-group/main.tf]: You could keep the existing code and just add the name line:
```
name = (var.create_iam_instance_profile ? aws_iam_instance_profile.this[0].arn : var.iam_instance_profile_arn) == null ? var.iam_role_name : null

Comment (sfbahr) [reply modules/self-managed-node-group/main.tf]: Only one of `arn` or `name` can be non-null: https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/launch_template#instance-profile

Comment (ichaban-pinterest) [reply modules/self-managed-node-group/main.tf]: Updated accordingly in https://github.com/pinterest/terraform-aws-eks/pull/3/commits/91e58f40aa16c408a07be5e8895ab8a59a88d93f

**Review by sfbahr — APPROVED** [review]
Looks good, thanks for addressing feedback
